### ISSUE TITLE
`npm unlink foo` doesn't reinstall the "real" `foo`

### ISSUE THREAD

**Issue by wearhere** [npm/npm#20995]
#### I'm opening this issue because:

npm is producing incorrect or undesirable behavior.

#### What's going wrong?

`npm unlink foo` doesn't reinstall the "real" `foo`. Even worse, it modifies my `package.json` to completely remove `foo`. To fix these issues, I have to do `npm unlink --no-save foo && npm install.`

This behavior is confusing because it means that `npm unlink` is not a mirror of `npm link`:

* `npm link ../foo` overwrites the "real" `foo` with the version of `foo` at `../foo`, but `npm unlink foo` doesn't replace `../foo` with the "real" `foo`
* `npm link ../foo` doesn't modify my `package.json` (to e.g. say `dependencies: { foo: "../foo" }`), but `npm unlink foo` modifies my `package.json` to remove `foo`

I realize that this behavior occurs because `npm unlink` is merely an alias for `npm uninstall`, as shown by running `npm unlink -h`. But I do not think it should be an alias, but rather its own command with its own behavior.

It seems like this is a partial design goal for you folks per https://github.com/npm/npm/issues/6248, "`npm link` and `npm unlink` should be mirror operations"… but I don't understand how you can consider them mirror operations if `npm unlink === npm uninstall`.

**TL;DR** `npm unlink foo` should made to behave like `npm unlink --no-save foo && npm install`, to make `npm unlink foo` the mirror of `npm link ../foo`.

#### How can the CLI team reproduce the problem?

As per the commands in the previous section, `s/foo/any-real-package`. I am glad to provide a test package if necessary.

### supporting information:

 - `npm -v` prints: 6.0.0
 - `node -v` prints: 8.9.3
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: OS X
 - Network issues:
   - Geographic location where npm was run: my local computer in an office in SF
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by wearhere** [comment]
Interestingly, it appears that if I do `npm unlink ../foo`, then the real module will be reinstated, and not removed from `package.json`, although npm may still make other modifications to `package.json` like alphabetizing the dependencies.

TL;DR `npm unlink ../foo` appears to be a shorter workaround than `npm unlink --no-save foo && npm install`.

**Comment by legodude17** [comment]
The problem is that `npm unlink` doesn't know if you replaced `foo` or added it.

**Comment by wearhere** [comment]
@legodude17 say more? If it's possible for `npm unlink --no-save foo && npm install` and `npm unlink ../foo` to do what I want I don't understand why `npm unlink foo` couldn't be revised to do the same thing.

**Comment by olets** [comment]
The issue's current title and some of what you've said assumes that foo was a project dependency before you ran `npm link foo`. If it wasn't, then `npm unlink foo` should not add it. Per legodude, the `npm unlink` command doesn't know whether your `npm link foo` replaced an existing `node_modules/foo[real]` with `node_modules/foo[link]`. I'd add that it could also have replaced an existing `node_modules/foo[brokenlink]` depending on what convolutions a person got up to.

But big 👍 to your bigger point, which stands without the `npm i foo && npm link foo && npm unlink foo` scenario (maybe edit the title?). For `npm link` and `npm unlink` to mirror each other, they should either both modify the project's dependency list or both not modify it. As it is now, the behavior of the one can't be predicted from the other.

---

I actually mostly come down on the other side: I'm surprised that `npm link` can't touch package.json — my expectation is `npm link` would be "npm add the linked version." With `--save | -S` and `--save-dev | -D` options, the result could be similar to `npm install username/repo`, adding e.g.

```
// result of `npm link foo`
"foo": "{prefix}/lib/node_modules/foo"
```

and overwriting `"foo": "semver"` if present. But then, sometimes your proposal feels the more intuitive direction to me too

**Comment by wearhere** [comment]
>  Per legodude, the npm unlink command doesn't know whether your npm link foo replaced an existing node_modules/foo[real] with node_modules/foo[link]

Pardon, but I still don't understand this. It seems clear to me that `npm unlink` _does_ know (or, at the least, _could know_, in my proposed behavior) from `package.json`:

* if `package.json` contains `foo` at the time that `npm unlink` is run, that means that `npm link foo` replaced an existing `node_modules/foo[real]`
* If `package.json` _doesn't_ contain `foo` at the time that `npm unlink` is run, that means that `npm link foo` did _not_ replace an existing `node_modules/foo[real]`.

(This works because `npm link` _doesn't_ touch `package.json`, to the second half of your message @olets. So I don't think we should change that behavior :)

--------------

You can verify what I say above like this:

* run `npm link ../foo` in a package that previously had `node_modules/foo[real]` installed. Then run `npm unlink --no-save foo && npm install`. You will notice that `node_modules/foo[real]` has been reinstalled.
* run `npm link ../foo` in a package that did _not_ previously have `node_modules/foo[real]` installed. Then run `npm unlink --no-save foo && npm install`. You will notice that `node_modules/foo[real]` does _not_ exist.

How did `npm install` "know whether `foo` was a project dependency or not before I ran `npm link ../foo`"? By checking `package.json`. If `npm install` can do this, then so could `npm unlink`. (And it seems that `npm unlink` may already do this if you unlink a local path, per [this comment](https://github.com/npm/npm/issues/20995#issuecomment-397459792)).

So it seems that npm is already fully capable of executing my desired behavior. I only want to execute that behavior when I run `npm unlink foo` vs. some other command.

**Comment by olets** [comment]
Thanks for the clarification 👍

Got it.

> `npm unlink foo` doesn't reinstall the "real" foo

sounds to me like a different behavior from

> `npm unlink foo` should made to behave like `npm unlink --no-save foo && npm install`

(I read the first as "`npm unlink foo` should behave like `npm unlink foo && npm install foo`"). But I see now you're saying it should reinstall the "real" version _if_ it was previously installed.

**Comment by wearhere** [comment]
You got it!
